app/processing/optimize/algos.py

```diff
@@ -0,0 +1,352 @@
+                self.uncomputes[new_reusable.raw] = True
+
+    @override
+    def compute(self) -> tuple[list[AncillaConnection], dict[ProgramNode, bool]]:
+        """Compute ancillas and uncomputes based on no-pred."""
+        while len(self.nopred) > 0:
+            self.current_node = self.pop_nopred()
+            self.nopred.extend(self.remove_node(self.current_node))
+
+            self.satisfy_dirty_qubit_requirement()
+            self.satisfy_reusable_qubit_requirement()
+
+            if len(self.current_node.qubit.returned_dirty_ids) > 0:
+                self.dirty.append(self.current_node)
+            if len(self.current_node.qubit.returned_reusable_ids) > 0:
+                self.reusable.append(self.current_node)
+            if (
+                len(
+                    self.current_node.qubit.returned_reusable_after_uncompute_ids,
+                )
+                > 0
+            ):
+                self.uncomputable.append(self.current_node)
+
+        return self.ancilla_edges, self.uncomputes
+
+
+class NoPredCheckNeedDiffScore(NoPred):
+    """NoPred variant with the check-need strategy + diff score.
+
+    Check need strategy:
+    All possible nodes are divided in one of two categories:
+    1. satisfied: the requirement of that node are satisfied by the available resources
+    2. unsatisfied: the requirement of that node can't be satisfied
+    We always prefer nodes that are in the first category.
+
+    Diff score:
+    Sort the nodes based on provided resources - required resources.
+
+    :param weight_reusable: weight of reusable qubits in diff-score
+    :param weight_uncomp: weight of uncomputable qubits in diff-score
+    :param weight_dirty: weight of dirty qubits in diff-score
+    """
+
+    weight_reusable: int = 2
+    weight_uncomp: int = 2
+    weight_dirty: int = 1
+
+    @override
+    def pop_nopred(self) -> ProcessedProgramNode:
+        total_reusable = sum(
+            [len(n.qubit.returned_reusable_ids) for n in self.reusable],
+        )
+        total_dirty = sum(
+            [len(n.qubit.required_dirty_ids) for n in self.dirty],
+        )
+        total_uncomputable = sum(
+            [
+                len(n.qubit.returned_reusable_after_uncompute_ids)
+                for n in self.uncomputable
+            ],
+        )
+        current_best: tuple[bool, int, None | ProcessedProgramNode] = (
+            False,
+            -maxsize - 1,
+            None,
+        )
+        for node in self.nopred:
+            required_dirty = len(node.qubit.required_dirty_ids)
+            required_reusable = len(node.qubit.required_reusable_ids)
+            satisfied = (
+                required_dirty < total_dirty + total_uncomputable
+                and required_reusable < total_reusable + total_uncomputable
+                and required_dirty + required_reusable
+                < total_dirty + total_reusable + total_uncomputable
```

Comment:
For me this looks clearer.
```suggestion
                and required_dirty + required_reusable < total_dirty + total_reusable + total_uncomputable
```